### PR TITLE
pysteps.utils with xarray

### LINKED ISSUE #198: `dB_transform` ignores existing metadata
Original report
----------------
(by Valentina Gregori)

> I also noticed another small issue (already in the previous version of pysteps). When I do a conversion to reflectivity using the function conversion.to_reflectivity(R, metadata) the output value of metadata['zerovalue'] is not as expected. I think this happens because in the transformation.dB_transform function, which is called by conversion.to_reflectivity, the code that checks if "zerovalue" is a key of metadata is missing:
> ```
> if zerovalue is None:
>     zerovalue = metadata.get("zerovalue", threshold-5)
> ```
> for now, it seems that the "zerovalue" is only replaced by "threshold-5". Do you think it makes sense?
> Thanks in advance!

### PR DESCRIPTION
**Pull request by dnerini**
Include new xarray-based data model into the `pysteps.utils` module (See #12).

## `utils.conversion`
- remove `metadata` arguments (since all attributes are now part of the input xr.DataArray object)
- make all methods available as [xarray accessors](http://xarray.pydata.org/en/stable/internals/extending-xarray.html) for DataArrays (e.g. can use both the `data_array.pysteps.to_rainrate()` or `pysteps.utils.to_rainrate(data_array)` syntax)
- `to_reflectivity()` has the option to convert to dB [dBZ] or keep linear scale [Z]
- add type hints for inputs and outputs
- reduce code duplication
- adapt tests

## `utils.dimension` (DEPRECATED)
- deprecate `aggregate_fields_time` in favor of xarray's `coarsen` method.
- deprecate `aggregate_fields_space` in favor of xarray's `coarsen` method.
- deprecate `aggregate_fields` in favor of xarray's `coarsen` method.
- deprecate `clip_domain` in favor of xarray's `sel` method.
- deprecate `square_domain` since it is obsolete.

## `utils.images`
- `morph_opening` takes xr.DataArray as input, and it also handles cases with time dimension "t"
- argument `n` of `morph_opening` is renamed `kernel` and can either be the size of the kernel or the kernel itself.
- update tests

## `utils.interpolate`
- all interpolation methods take and return xarray's object (both Dataset and DataArray)
- migrate deprecated scipy's `Rbf` to new `RBFInterpolator` (scipy>=1.7)
- update tests

## `utils.transformation`
- remove `metadata` arguments (since all attributes are now part of the input xr.DataArray object)
- make all methods available as [xarray accessors](http://xarray.pydata.org/en/stable/internals/extending-xarray.html) for DataArrays (e.g. can use both the `data_array.pysteps.sqrt_transform()` or `pysteps.utils.sqrt_transform(data_array)` syntax)
- add general `back_transform` method to remove any existing data transformation.
- `boxcox_transform` and `db_transform` use a simpler offset to avoid computing log(0), instead of having to define the `threshold` and `zerovalue` parameters as in V1 (closes #198)
- refactor `nq_transform`: in particular, the inverse transformation needs an empirical target distribution (the `template` argument).
- add type hints for inputs and outputs
- reduce code duplication
- rewrite tests

### CODE REVIEW

**Comment by dnerini** [comment]
OK let's say this is ready for review!

- still failing tests (need to adapt all methods that use the utils module)
- didn't touch at the `utils.spectral` module for now, do we want to adopt xarray for complex data too? this would be similar to what is done in [xrft](https://xrft.readthedocs.io/en/latest/). Proposition: should we adopt xrft ? @pulkkins 

I know it is a lot of changes, I don't expect you to go through all of them. Rather, please try to see whether you agree with the general changes in the interfaces listed in the summary at the top.

Thanks!

**Comment by RubenImhoff** [comment]
This is an amazing amount of work, @dnerini! I'll try to have a look at it soon.

File: examples/anvil_nowcast.py
Comment (RubenImhoff) [inline examples/anvil_nowcast.py]: Very minor: I believe we have: 
`# -*- coding: utf-8 -*-`
in most of the scripts

File: examples/anvil_nowcast.py
Comment (RubenImhoff) [inline examples/anvil_nowcast.py:90]: No need to define the threshold and zerovalue, or are the same values present in the "metadata" of the DataArray?

Comment (RubenImhoff) [reply examples/anvil_nowcast.py:90]: (it will work anyway, just checking that the numbers are the same)

File: pysteps/io/exporters.py
Comment (RubenImhoff) [inline pysteps/io/exporters.py]: Similar to var_xc?
```suggestion
    var_yc.units = metadata.get("cartesian_unit", "m")
```

File: pysteps/io/nowcast_importers.py
Comment (RubenImhoff) [inline pysteps/io/nowcast_importers.py:110]: What will we do with the quality field? It is useful, but so far, it is not used to my knowledge.

File: pysteps/tests/test_io_mrms_grib.py
Comment (RubenImhoff) [inline pysteps/tests/test_io_mrms_grib.py:36]: I like the implementation of the bounding_box!

File: pysteps/utils/conversion.py
Comment (RubenImhoff) [inline pysteps/utils/conversion.py:38]: Is this something we still want to set in the attributes of the DataArray?

File: pysteps/utils/interface.py
Comment (RubenImhoff) [inline pysteps/utils/interface.py]: Seems like:
`# -*- coding: utf-8 -*-`
is twice in there.

**Review by RubenImhoff — APPROVED** [review]
Amazing work, @dnerini! Because you've done so much work, I can't promise that I've spotted all possible mistakes in the code, so let's also wait for another review by one of the others. ;)

File: examples/anvil_nowcast.py
Comment (dnerini) [reply examples/anvil_nowcast.py:90]: The idea is to replace them with a simpler offset, see for example [here](https://github.com/pySTEPS/pysteps/blob/c7a9e26f829334c2eb657be2591fd4100ef09044/pysteps/utils/transformation.py#L142).

This is not yet enforced in all methods, especially in the examples, which will need to be updated along as we refactor the pysteps methods.

**Pull request merged**
